### PR TITLE
A pile of misc doc fixes.

### PR DESCRIPTION
**Pull request by ezyang**
CC @apaszke, who was the source of many of these clarifications.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: torch/csrc/autograd/python_engine.cpp [hunk 1]
```diff
@@ -97,13 +97,33 @@ void compute_partial_exec_callbacks(const function_list& roots,
...
+//      If false, the graph used to compute the grad will be freed.
+//  inputs (tuple of Variable)
+//      Inputs w.r.t. which the gradient will be returned.  This
+//      function will return a tuple containing an entry for each
+//      input variable.  Non-mentioned leaf variables will have
+//      gradients accumulated to grad, unless only_inputs below
+//      is True.
+//  only_inputs (boolean)
+//      If false, any inputs not specified in inputs will have their
+//      gradients accumulated to grad.
```
Comment (apaszke) [inline torch/csrc/autograd/python_engine.cpp]: I don't it's a good idea to replicate Python docs here. You can always look them up, and they'll quickly get out of date

File: torch/csrc/autograd/python_engine.cpp [hunk 2]
```diff
@@ -137,6 +157,11 @@ PyObject *THPEngine_run_backward(THPEngine *self, PyObject *args, PyObject *kwar
     auto& variable = ((THPVariable*)_variable)->cdata;
     THPUtils_assert(!variable->is_volatile,
         "element %d of variables tuple is volatile", i);
+    // If grad_fn is NULL (as is the case for a leaf node), we instead
+    // interpret the gradient function to be a grad accumulator,
+    // which will accumulate its inputs into the grad property of the
+    // variable.  If inputs/only_inputs is specified, a callback is
+    // used to suppress the operation of this node.
```
Comment (apaszke) [inline torch/csrc/autograd/python_engine.cpp]: the callback comment doesn't refer to this part of code, so it's not the best place for it

File: torch/csrc/autograd/python_engine.cpp [hunk 3]
```diff
@@ -156,7 +181,7 @@ PyObject *THPEngine_run_backward(THPEngine *self, PyObject *args, PyObject *kwar
   Engine::callback_map callbacks;
   CallbackContext ctx;
   if (inputs != NULL) {
-    THPUtils_assert(PyTuple_Check(inputs), "outputs argument has to be a tuple");
+    THPUtils_assert(PyTuple_Check(inputs), "inputs argument has to be a tuple");
```
Comment (apaszke) [inline torch/csrc/autograd/python_engine.cpp]: I think it's `outputs` in Python

**Comment by fmassa** [comment]
I think it would also be helpful to mention that [`next_functions`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_cpp_function.cpp#L93) always returns a new python function.
So once can't rely on the address of functions defined in C like (like convolution and BN).
```python
m = nn.Conv2d(1, 1, 1)
out = m(Variable(torch.rand(1, 1, 3, 3)))
print(out.grad_fn)
print(out.grad_fn) # different ref
```

**Comment by apaszke** [comment]
@fmassa I don't think it's a good idea, because it's not an invariant that we'd like to maintain. In fact, it'd be best to get rid of it, because it causes #1626.

File: torch/csrc/autograd/python_engine.cpp (hunk 3 above)
Comment (ezyang) [inline torch/csrc/autograd/python_engine.cpp]: It's `inputs` in Python. Here is the header of this function:

```
  const char *accepted_kwargs[] = {"variables", "grad_variables",
      "keep_graph", "inputs", "only_inputs", NULL}; 
  if (!PyArg_ParseTupleAndKeywords(args, kwargs, "OOb|Ob", (char**)accepted_kwargs,
        &variables, &grad_variables, &keep_graph, &inputs, &only_inputs))
```

**Comment by ezyang** [comment]
I can't imagine someone relying on grad_fn returning the same Python object; usually it's the opposite. A warning might still be helpful. Up to you.

**Comment by fmassa** [comment]
@ezyang me neither, but it has bitten me this weekend, so I thought that it would be helpful to mention it. But if you are fixing it, even better! :-)

**Pull request merged**